Pyic 8253 - Stored Identity Service Api Tests

undermad opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

API tests for stored identity service

### Why did it change

Storing a reusable identity allows returning users to access future services without re-proving their identity.  
By capturing the most recent VC bundle and identity state after key journeys, the system ensures that the SIS record remains accurate, valid, and aligned with the latest user information.  

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-8253](https://govukverify.atlassian.net/browse/PYIC-8253)

## Checklists

<!-- Delete if changes don't apply -->
- [x] API/unit/contract tests have been written/updated


[PYIC-8253]: https://govukverify.atlassian.net/browse/PYIC-8253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

undermad committed:
PYIC-8253: Add api tests for stored identity feature - fraud checks and reuse existing identity

undermad committed:
PYIC-8253: Add api tests for stored identity feature - M1C outcomes

undermad committed:
PYIC-8253: Add additional check in api test to make sure stored identity record wasn't created.

undermad committed:
PYIC-8253: Add api tests for stored identity service against D02 journeys.

undermad committed:
PYIC-8253: Add api tests for stored identity service against V03 journeys.

thebauSoftwire commented on api-tests/features/cimit/p1-alternate-doc.feature:
```diff
@@ -32,6 +32,7 @@ Feature: P1 CIMIT - Alternate doc
       Then I get an OAuth response
       When I use the OAuth response to get my identity
       Then I get a 'P1' identity
+      And I don't have a stored identity in EVCS
```
I think it's fine to just add this for one or two key journeys e.g. a new identity journey since we're just checking that if the feature flag is turned off then we don't store an identity. It'll also make it easier when we turn on the feature flag as we won't have to update all of these tests

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,247 @@
+        | kenneth-driving-permit-valid |
+
+  Rule: P1 - V03 Enhanced Verification
+    Scenario Outline: Same session F2F enhanced verification mitigation
```
I think it might be good to also have separate session V03 mitigations - we could check that the stored identity wasn't created before they start a new session. To minimise the number of tests, we could just have one P1 test for this

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,247 @@
+Feature: Stored Identity Service - CIMIT journeys
+  Rule: P1 - D02 Mitigation
+    Background:
+      Given I activate the 'storedIdentityService' feature set
```
this could be moved to a top-level background section above all the rules to turn on the feature set

thebauSoftwire commented on api-tests/features/stored-identity/m1c-journeys.feature:
```diff
@@ -0,0 +1,142 @@
+      When I submit an 'appTriage' event
+      Then I get a 'dcmaw' CRI response
+
+    Scenario: Successful M1C P2 identity via DCMAW using chipped passport
+      When I submit 'kenneth-passport-valid' details to the CRI stub
+      Then I get a 'page-dcmaw-success' page response
+      When I submit a 'next' event
+      Then I get an 'address' CRI response
+      When I submit 'kenneth-current' details to the CRI stub
+      Then I get a 'fraud' CRI response
+      When I submit 'kenneth-unavailable' details to the CRI stub
+      Then I get a 'page-ipv-success' page response
+      When I submit a 'next' event
+      Then I get an OAuth response
+      When I use the OAuth response to get my identity
+      Then I get a 'P2' identity
+      And I have a 'GPG45' stored identity record type with a 'P2' vot
```
I think this is a duplicate of [this one in /stored-identity/p2-journeys.feature](https://github.com/govuk-one-login/ipv-core-back/blob/09fa4cba28d68b97705aa9da10a223e9bfdbb269/api-tests/features/stored-identity/p2-journeys.feature#L117-L132)? I think it might also be fine to just have that or just this test for new M1C profile (uk address) so we don't have too many duplicate tests

undermad replied:
I deleted test from `p2-journeys.feature` and also moved another `No stored identity - unsuccessful M1C journey` to `m1c-journeys.feature` to keep all related tests in one place

thebauSoftwire commented on api-tests/features/stored-identity/m1c-journeys.feature:
```diff
@@ -0,0 +1,142 @@
+      Then I get an OAuth response
+      When I use the OAuth response to get my identity
+      Then I get a 'P2' identity
+      And I have a 'GPG45' stored identity record type with a 'P2' vot
```
To reduce duplicate tests it might also be fine to just have `Scenario Outline: Existing M1C address and name change - <selected-name-change>` or `Scenario Outline: Existing M1C name change - <selected-name-change>` and `Scenario: Existing M1C address change` especially as you've also added more update scenarios in reuse and rfc files :)

thebauSoftwire commented on api-tests/features/stored-identity/p1-reuse-existing-identity.feature:
```diff
@@ -0,0 +1,85 @@
+      | family-name-only     | kenneth-changed-family-name-driving-permit-valid | kenneth-changed-family-name-score-2 | Kenneth             | Smith                |
+      | given-names-only     | kenneth-changed-given-name-driving-permit-valid  | kenneth-changed-given-name-score-2  | Ken                 | Decerqueira          |
+
+  Scenario Outline: Successful Address and Name Change - <selected-name-change>
```
we could probably also remove this as it's very similar to the above two

thebauSoftwire commented on api-tests/features/stored-identity/p1-repat-fraud-check.feature:
```diff
@@ -0,0 +1,94 @@
+      | given-names-only     | kenneth-changed-family-name-driving-permit-valid | kenneth-changed-family-name-score-2 | Kenneth             | Smith                |
+      | family-name-only     | kenneth-changed-given-name-driving-permit-valid  | kenneth-changed-given-name-score-2  | Ken                 | Decerqueira          |
+
+  Scenario Outline: Fraud 6 Months Expiry + Address and Name Change - <selected-name-change>
```
again, might be fine to just have this test or the ones above to reduce test duplicates

thebauSoftwire reviewed (commented):
Nice test coverage! Just a couple of comments to potentially reduce test duplication

undermad committed:
PYIC-8253: Refactor stored identity service api tests.

Tests refactor was required according to code review, removes duplicate codes and add additional test coverage.

undermad committed:
PYIC-8253: Add reprove api test for stored identity service

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,236 @@
+      Examples:
+        | document-details             |
+        | kenneth-passport-valid       |
+        | kenneth-driving-permit-valid |
```
this is technically a same-session mitigation journey as the next step on line 188 is starting the mitigation journey on the same session they received the CI :)

undermad replied:
Ah, my bad - it should be "same session" as separate session is below

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,236 @@
+        | kenneth-driving-permit-valid |
+
+    Scenario Outline: Successful F2F enhanced verification mitigation - separate session
+      And I don't have a stored identity in EVCS
```
I don't think we need this line as we have the top-level background step on line 4

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,236 @@
+      Examples:
+        | document-details             |
+        | kenneth-passport-valid       |
+        | kenneth-driving-permit-valid |
```
it's probably fine to also just do one of these cases and replace this `Scenario Outline` with just a normal `Scenario`

thebauSoftwire commented on api-tests/features/stored-identity/cimit-journeys.feature:
```diff
@@ -0,0 +1,236 @@
+      Examples:
+        | document-details             |
+        | kenneth-passport-valid       |
+        | kenneth-driving-permit-valid |
```
same here, probably fine to do just one of these cases

thebauSoftwire commented on api-tests/features/stored-identity/p1-journeys.feature:
```diff
@@ -1,3 +1,4 @@
+@Build
```
nice!

thebauSoftwire reviewed (commented):
Looking good! Just a couple of small comments

undermad committed:
PYIC-8253: Refactor cimit-journeys api tests

Additional refactoring was required according to code review.
